### PR TITLE
Add tests for Qobj

### LINKED ISSUE #653: Add more tests about the `QObj` themselves
### What is the expected enhancement?
Test the `Qobj` class itself if some crucial logic is implemented there.

### PR DESCRIPTION
**Pull request by Max-Reuter**
Fixes #653.
Partially implement #651 as a unit test.

### Summary
Add tests for instantiating Qobj elements and dictionary functionality and JSON validation of Qobj.

### CODE REVIEW

File: test/python/test_qobj.py [hunk 1]
```diff
@@ -28,6 +32,24 @@ def test_create_qobj(self):
...
+            header={},
+            config=QobjConfig(shots=1024, memory_slots=2, max_credits=10),
+            experiments=[
+                QobjExperiment(instructions=[
+                    QobjInstruction(name='u1', qubits=[1], params=[0.4]),
+                    QobjInstruction(name='u2', qubits=[1], params=[0.4, 0.2])
+                ])
+            ]
+        )
+        qobj._version = '67890'  # private member variables shouldn't appear in the dict
```
[the remark this answers is not in the document]
Comment (Max-Reuter) [reply test/python/test_qobj.py]: I've kept a single test of this in ```test_as_dict``` because it can't be independently tested due to Qobj requiring certain args upon initialization; thus, as_dict will always contain more information than just the ```_version```.

File: test/python/test_qobj.py [hunk 2]
```diff
@@ -39,7 +40,82 @@ def test_create_qobj(self):
...
+        shots = 1
+        memory_slots = 2
+
+        qobj_config = QobjConfig(shots=shots, memory_slots=memory_slots)
+
+        self.assertTrue(
+            all(getattr(qobj_config, required_arg) is not None
+                for required_arg in QobjConfig.REQUIRED_ARGS))
+
+        with self.assertRaises(ValueError):
```
Comment (delapuente) [inline test/python/test_qobj.py]: It is not implemented for a `Qobj` to raise in case there is a required argument missing. As commented offline, talk to @diego-plan9 about adding this functionality to the class. What is sure is that `from_dict` behaves in a very similar way as your test expect. Add tests for `from_dict` based on this one.

Comment (Max-Reuter) [reply test/python/test_qobj.py]: @diego-plan9 and I decided that TypeErrors should be raised if the required arguments are not provided, as Python does by default. The assertions that the errors are raised helps validate that the errors haven't been circumvented somehow, such as being intercepted by a try/except block. ```from_dict``` is in the process of being removed so I won't add tests based on this one.

File: test/python/test_qobj.py [hunk 3]
```diff
@@ -42,48 +51,29 @@ def setUp(self):
...
-    def test_as_dict_to_json(self):
+    def test_as_dict_against_schema(self):
         """Test dictionary representation of Qobj against its schema."""
         file_path = self._get_resource_path('qobj_schema.json', Path.SCHEMAS)
 
         with open(file_path, 'r') as schema_file:
             schema = json.load(schema_file)
 
-        jsonschema.validate(self.valid_qobj.as_dict(), schema)
+        passed = False
```
Comment (delapuente) [inline test/python/test_qobj.py]: Remember when we talked about how `validate()` worked? We assumed it was returning a boolean indicating if the validation was successful or not but it turns out it raises an exception! Your code is OK but you are reimplementing some Python `unittest` functionality such as printing the reason for which the test is failing. You can simply do:

```python
try:
  jsonschema.validate(self._valid_qobj.as_dict(), schema)
except jsonschema.ValidationError as validation_error:
  self.fail(str(validation_error))
```

This is slightly different to the first version of your test where you have the `validate()` line only. If this version enters the `except` block it will **fail**. With your first version, if the line throws, the test result is **error**.

File: test/python/test_qobj.py (hunk 2 above)
Comment (diego-plan9) [reply test/python/test_qobj.py]: Elaborating a bit - I basically see arguments for any of the three options:
* removing the `with self.assertRaises` block - technically, we are testing Python functionality (ie. a function, in this case the constructor, needs the positional parameters), which is a given.
* keeping the `with self.assertRaises` block - with this, we are making explicit that we satisfy the original requirement ("a QobjFoo should have bar, baz as required attributes"), decoupled from the current implementation.
* changing the test so it uses `from_dict`  instead of instantation: `from_dict`, even if it is going to be hopefully short-lived, is also responsible from checking that  the attributes are present.

So basically whichever @delapuente and @Max-Reuter agree upon sounds good to me!

Comment (delapuente) [reply test/python/test_qobj.py]: I did not realize mandatory arguments are modelled as positional arguments without defaults so there is no need for extra validation. My only concern is passing `None` which would trick Python into generating an invalid dictionary representation. But let's move that effort to #654.

Although it's true that `from_dict` is on its way of being removed, the functionality won't disappear (see #652) so these test will remain valid.

So, I would recommend creating tests for `from_dict` and testing that validation works there. @Max-Reuter what do you think?

Comment (Max-Reuter) [reply test/python/test_qobj.py]: I propose we remove the ```assertRaises``` block altogether and rely on lines 84-85 to handle the ```None``` case, if we don’t need to check that the error is being raised and not circumvented. I don’t see the use of ```from_dict``` if all we’re doing is checking the arguments aren’t ```None```. Python will handle the case where the arguments aren’t passed at all.

File: test/python/test_qobj.py [hunk 4]
```diff
@@ -39,7 +40,46 @@ def test_create_qobj(self):
...
+            instructions=[QobjInstruction(name='u1', qubits=[1], params=[0.4])]
+        )
+        valid_qobj_instruction = QobjInstruction(name='u1', qubits=[1], params=[0.4])
+
+        dicts = [self.valid_dict, qobj_config_dict, qobj_experiment_dict, qobj_instruction_dict]
+        objects = [self.valid_qobj, valid_qobj_config,
+                   valid_qobj_experiment, valid_qobj_instruction]
+        classes = [Qobj, QobjConfig, QobjExperiment, QobjInstruction]
+
+        for objekt, dikt, klass in zip(objects, dicts, classes):
```
Comment (diego-plan9) [inline test/python/test_qobj.py]: Small nit - can you revise the names of these variables? From [PEP 8](https://www.python.org/dev/peps/pep-0008/#function-and-method-arguments):

> If a function argument's name clashes with a reserved keyword, it is generally better to append a single trailing underscore rather than use an abbreviation or spelling corruption. Thus class_ is better than clss

That paragraph is related to function argument's names, but applies in general too.

Comment (delapuente) [reply test/python/test_qobj.py]: I completely agree with Diego but try to come up with better names. You have `Qobj` instances (not `object` instances), `QItem` derived classes and expected dictionaries. If you can not come up with meaningful names, adhere to Diego's recommendation.

File: test/python/test_qobj.py [hunk 5]
```diff
@@ -60,38 +55,31 @@ def test_as_dict_against_schema(self):
...
+        valid_qobj_config = QobjConfig(shots=1024, memory_slots=2, max_credits=10)
+        valid_qobj_experiment = QobjExperiment(
+            instructions=[QobjInstruction(name='u1', qubits=[1], params=[0.4])]
+        )
+        valid_qobj_instruction = QobjInstruction(name='u1', qubits=[1], params=[0.4])
 
+        dicts = [self.valid_dict, qobj_config_dict, qobj_experiment_dict, qobj_instruction_dict]
+        objects = [self.valid_qobj, valid_qobj_config,
+                   valid_qobj_experiment, valid_qobj_instruction]
+        classes = [Qobj, QobjConfig, QobjExperiment, QobjInstruction]
```
Comment (delapuente) [inline test/python/test_qobj.py]: Non-blocking but a recommendation: if you create a more explicit data structure, you won’t need to zip it after, you increase the readability of the test (avoiding the overhead introduced by many variables), and you won’t rely on indices to keep related parameters together:

```python
test_parameters = {
  QobjConfig: (
    QobjConfig(shots=1024, memory_slots=2, max_credits=10),
    {'shots': 1, 'memory_slots': 2}
  ),
  ...
}
```

Then you can iterate it by doing:

```python
for class_, (qobj, expected_dict) in test_parameters.items():
  with self.subTest(...):
    # test things
```

File: test/python/test_qobj.py (hunk 4 above)
Comment (delapuente) [reply test/python/test_qobj.py]: I completely agree with Diego but try to come up with better names. You have `Qobj` instances (not `object` instances), `QItem` derived classes and expected dictionaries. If you can not come up with meaningful names, adhere to Diego's recommendation.

File: test/python/test_qobj.py [hunk 6]
```diff
@@ -39,7 +40,46 @@ def test_create_qobj(self):
...
+
+        with open(file_path, 'r') as schema_file:
+            schema = json.load(schema_file)
+
+        try:
+            jsonschema.validate(self.valid_qobj.as_dict(), schema)
+        except jsonschema.ValidationError as validation_error:
+            self.fail(str(validation_error))
+
+    @unittest.expectedFailure  # expected to fail until _qobjectify_item is updated
```
Comment (delapuente) [inline test/python/test_qobj.py]: This is very vague, reference the line with the `TODO` you found.
